sites/washington_city_paper.py

```diff
@@ -50,134 +50,125 @@
+        ]
+
+        for prop, func in scraper_funcs:
+            try:
+                val = func(page, soup)
+            except Exception as e:
+                raise ArticleScrapingError(
+                    ScrapeFailure.MALFORMED_RESPONSE, path, external_id, f"Error scraping {prop} for article path"
+                ) from e
+            metadata[prop] = val
+
+        return metadata
+
+    def fetch_article(self, external_id: str, path: str) -> Response:
+        url = f"https://{DOMAIN}{path}"
 
-    for prop, func in scraper_funcs:
         try:
-            val = func(page, soup)
+            page = safe_get(url)
         except Exception as e:
             raise ArticleScrapingError(
-                ScrapeFailure.MALFORMED_RESPONSE, path, external_id, f"Error scraping {prop} for article path"
+                ScrapeFailure.FETCH_ERROR, path, str(external_id), f"Request failed for {url}"
             ) from e
-        metadata[prop] = val
-
-    return metadata
-
-
-def validate_not_excluded(page: Response) -> Optional[str]:
-    soup = BeautifulSoup(page.text, features="html.parser")
-    primary = soup.find(id="primary")
-
-    if primary:
-        classes = {value for element in primary.find_all(class_=True) for value in element["class"]}
-        if "tag-exclude" in classes:
-            return "Article has exclude tag"
-
-    return None
-
-
-def fetch_article(external_id: str, path: str) -> Response:
-    url = f"https://{DOMAIN}{path}"
-
-    try:
-        page = safe_get(url)
-    except Exception as e:
-        raise ArticleScrapingError(ScrapeFailure.FETCH_ERROR, path, str(external_id), f"Request failed for {url}") from e
-
-    error_msg = validate_response(page, [validate_status_code, validate_not_excluded])
-    if error_msg is not None:
-        raise ArticleScrapingError(ScrapeFailure.FAILED_SITE_VALIDATION, path, str(external_id), error_msg)
-
-    return page
-
 
-WCP_SITE = Site(
-    NAME,
-    FIELDS,
-    TRAINING_PARAMS,
-    SCRAPE_CONFIG,
-    transform_raw_data,
-    extract_external_id,
-    scrape_article_metadata,
-    fetch_article,
-    bulk_fetch,
-    bulk_fetch_by_external_id,
-    get_article_text,
-    POPULARITY_WINDOW,
-    MAX_ARTICLE_AGE,
+        error_msg = validate_response(page, [validate_status_code, self.validate_not_excluded])
+        if error_msg is not None:
+            raise ArticleScrapingError(ScrapeFailure.FAILED_SITE_VALIDATION, path, str(external_id), error_msg)
+
+        return page
+
+    def bulk_fetch_by_external_id(self, external_ids) -> None:
+        # https://stackoverflow.com/questions/16706956/is-there-a-difference-between-raise-exception-and-raise-exception-without
+        raise NotImplementedError
+
+    @staticmethod
+    def get_article_text(metadata: Dict[str, Any]) -> None:
+        raise NotImplementedError
+
+    def bulk_fetch(self, start_date: date, end_date: date) -> List[Dict[str, Any]]:
+        raise NotImplementedError
+
+    @staticmethod
```

Comment:
🤔 What's the motivation behind setting this method and other ones below it static? Is it to signify site-specific utilities that aren't part of what the `Site` abstract class requires?

(The same question applies to similar methods in the other two site classes as well.)

Reply:
Nice! In that case, should we make `transform_raw_data` a static method too, since it's not using `self` in any of the three site implementations?

Reply:
Yes, we should 🙂 good catch